src/NServiceBus.AzureStoragePersistence/Subscriptions/Azure/AzureSubscriptionStorage.cs

```diff
@@ -70,62 +82,49 @@ void ISubscriptionStorage.Subscribe(Address address, IEnumerable<MessageType> me
+                select s;
+
+            var subscription = query.AsTableQuery().AsEnumerable().SafeFirstOrDefault();
+            if (subscription != null)
             {
-                var query = from s in table.CreateQuery<Subscription>()
-                    where s.PartitionKey == messageType.ToString() && s.RowKey == encodedAddress
-                    select s;
-                var subscription = query.AsTableQuery().AsEnumerable().SafeFirstOrDefault();
-                if (subscription != null)
-                {
-                    var operation = TableOperation.Delete(subscription);
-                    table.Execute(operation);
-                }
+                var operation = TableOperation.Delete(subscription);
+                return table.ExecuteAsync(operation);
             }
+
+            return TaskEx.CompletedTask;
         }
 
         /// <summary>
         /// Returns the subscription address based on message type
         /// </summary>
-        /// <param name="messageTypes">Types of messages that subscription addresses should be found for</param>
+        /// <param name="messageTypes">Types of messages that subscription addresses should sbe found for</param>        
+        /// <param name="context">The current pipeline context</param>
         /// <returns>Subscription addresses that were found for the provided messageTypes</returns>
-        IEnumerable<Address> ISubscriptionStorage.GetSubscriberAddressesForMessage(IEnumerable<MessageType> messageTypes)
+        public Task<IEnumerable<Subscriber>> GetSubscriberAddressesForMessage(IEnumerable<MessageType> messageTypes, ContextBag context)
         {
-            var subscribers = new List<Address>();
+            var subscribers = new List<Subscriber>();
             var table = client.GetTableReference(subscriptionTableName);
 
             foreach (var messageType in messageTypes)
             {
                 var query = from s in table.CreateQuery<Subscription>()
-                    where s.PartitionKey == messageType.ToString()
-                    select s;
-
-                subscribers.AddRange(query.ToList().Select(s => Address.Parse(DecodeFrom64(s.RowKey))));
+                            where s.PartitionKey == messageType.ToString()
+                            select new Subscriber(DecodeFrom64(s.RowKey), new EndpointName(s.EndpointName));
+                
+                subscribers.AddRange(query.ToList());
             }
 
-            return subscribers;
-        }
-
-        public void Init()
-        {
-            //No-op
-        }
-
-        static string EncodeTo64(string toEncode)
-        {
-            return Convert.ToBase64String(Encoding.ASCII.GetBytes(toEncode));
-        }
-
-        static string DecodeFrom64(string encodedData)
-        {
-            return Encoding.ASCII.GetString(Convert.FromBase64String(encodedData));
+            return Task.FromResult((IEnumerable<Subscriber>)subscribers);
```

Comment:
this doesn't need to be cast to `IEnumerable<Subscriber>`. You can do `Task.FromResult<IEnumerableSubscriber>(subscribers)`